File: Algorithm/Python/1000+/2497. Maximum Star Sum of a Graph.py
from collections import defaultdict
import heapq

class Solution:
    def maxStarSum(self, vals: List[int], edges: List[List[int]], k: int) -> int:
        neighbors = defaultdict(list)
        for s, e in edges:
            neighbors[s].append(vals[e])
            neighbors[e].append(vals[s])
        
        result = max(vals)
        for center, neighbor in neighbors.items():
            k_neighbors = heapq.nlargest(k, neighbor)
            # 写法一
            # result = max(result, vals[center]+sum(num if num > 0 else 0 for num in k_neighbors))
            # 写法二
            result = max(result, vals[center]+sum(num for num in k_neighbors if num > 0))
        return result

# 可以在构建图的时候就把 < 0 的 neighbor 卡掉，这样时间复杂度会更小
class Solution:
    def maxStarSum(self, vals: List[int], edges: List[List[int]], k: int) -> int:
        neighbors = defaultdict(list)
        for s, e in edges:
            if vals[e] > 0:
                neighbors[s].append(vals[e])
            if vals[s] > 0:
                neighbors[e].append(vals[s])
        
        result = max(vals)
        for center, neighbor in neighbors.items():
            result = max(result, vals[center]+sum(heapq.nlargest(k, neighbor)))
        return result

File: Algorithm/C++/1000+/2497. Maximum Star Sum of a Graph.cc
#include <queue>
#include <vector>
#include <unordered_map>
#include <algorithm> 

class Solution {
public:
    int maxStarSum(std::vector<int>& vals, std::vector<std::vector<int>>& edges, int k) {
        std::unordered_map<int, std::vector<int>> neighbors;
        for (auto& edge: edges) {
            if (vals[edge[1]] > 0) {
                neighbors[edge[0]].push_back(vals[edge[1]]);
            }
            if (vals[edge[0]] > 0) {
                neighbors[edge[1]].push_back(vals[edge[0]]);
            }
        }

        int result = *std::max_element(vals.begin(), vals.end());
        for(auto& _pair: neighbors) {
            std::priority_queue pq(_pair.second.begin(), _pair.second.end()); // 默认是 max heap
            int sum=vals[_pair.first], count = 0;
            while(!pq.empty() && count < k) {
                sum += pq.top();
                pq.pop();
                count++;
            }
            result = std::max(result, sum);
        }
        return result;
    }
};